examples/c_api/aggregates_string.c

```diff
@@ -0,0 +1,365 @@
+  TRY(ctx, tiledb_query_submit(ctx, query));
+
+  // Prepare data for second write
+  char coords_rows_2[] = {"garplygraultgubquux"};
+  uint64_t coords_rows_size_2 = sizeof(coords_rows_2);
+  uint64_t coords_rows_offsets_2[] = {0, 6, 12, 15};
+  uint64_t coords_rows_offsets_size_2 = sizeof(coords_rows_offsets_2);
+  int coords_cols_2[] = {1, 2, 3, 4};
+  uint64_t coords_cols_size_2 = sizeof(coords_cols_2);
+  int data_2[] = {6, 6, 3, 4};
+  uint64_t data_size_2 = sizeof(data_2);
+
+  // Reset buffers
+  TRY(ctx, tiledb_query_set_data_buffer(ctx, query, "a", data_2, &data_size_2));
+  TRY(ctx,
+      tiledb_query_set_data_buffer(
+          ctx, query, "rows", coords_rows_2, &coords_rows_size_2));
+  TRY(ctx,
+      tiledb_query_set_offsets_buffer(
+          ctx,
+          query,
+          "rows",
+          coords_rows_offsets_2,
+          &coords_rows_offsets_size_2));
+  TRY(ctx,
+      tiledb_query_set_data_buffer(
+          ctx, query, "cols", coords_cols_2, &coords_cols_size_2));
+
+  // Submit second query
+  TRY(ctx, tiledb_query_submit(ctx, query));
+
+  // Finalize query (IMPORTANT)
+  TRY(ctx, tiledb_query_finalize(ctx, query));
+
+  // Close array
+  tiledb_array_close(ctx, array);
+
+  // Clean up
+  tiledb_array_free(&array);
+  tiledb_query_free(&query);
+  tiledb_ctx_free(&ctx);
+}
+
+void print_cells(
+    uint64_t result_num,
+    uint64_t* rows_offsets,
+    uint64_t rows_data_size,
+    char* rows_data,
+    int32_t* cols_data,
+    int32_t* a_data) {
+  for (uint64_t r = 0; r < result_num; r++) {
+    // For strings we must compute the length based on the offsets
+    uint64_t row_start = rows_offsets[r];
+    uint64_t row_end =
+        r == result_num - 1 ? rows_data_size : rows_offsets[r + 1] - 1;
+    const int row_value_size = row_end - row_start + 1;
+    const char* row_value = &rows_data[row_start];
+
+    const int32_t col_value = cols_data[r];
+    const int32_t a_value = a_data[r];
+    printf(
+        "Cell (%.*s, %i) has data %d\n",
+        row_value_size,
+        row_value,
+        col_value,
+        a_value);
+  }
+}
+
+void read_array() {
+  // Create TileDB context
+  tiledb_ctx_t* ctx;
+  tiledb_ctx_alloc(NULL, &ctx);
+
+  // Open array for reading
+  tiledb_array_t* array;
+  tiledb_array_alloc(ctx, array_name, &array);
+  tiledb_array_open(ctx, array, TILEDB_READ);
+
+  // Calculate maximum buffer sizes
+  uint64_t max_size = 64;  // variable-length result has unknown size
+  uint64_t max_offsets_size = sizeof(uint64_t);
+  uint64_t min_size = 64;  // variable-length result has unknown size
+  uint64_t min_offsets_size = sizeof(uint64_t);
+
+  // Aggregate result buffers (1 cell each of unknown size)
+  char* max = (char*)malloc(max_size);
+  uint64_t max_offsets[1];
+  char* min = (char*)malloc(min_size);
+  uint64_t min_offsets[1];
+
+  // Attribute/dimension buffers
+  // (unknown number of cells, buffer sizes are estimates)
+  const size_t NUM_CELLS = 2;
+  char rows_data[NUM_CELLS * 16];
```

Comment:
MSVC does not like the use of `NUM_CELLS` here. Declaring it as `constexpr` would help, but it was introduced in C23.

Reply:
d110916